Earthfile

```diff
@@ -0,0 +1,216 @@
+    # set locale to utf-8, which is required for some odbc drivers (mysql);
+    ENV LC_ALL=C.UTF-8
+
+    RUN apt-get update && apt-get upgrade -y && \
+        apt-get install -y build-essential zlib1g-dev \
+        wget unixodbc cmake g++ \
+        odbc-postgresql postgresql-client gnupg apt-transport-https && \
+        apt-get clean
+
+    # we need an mysql odbc driver for the integration tests
+    # currently the test fail with a newer driver
+    RUN cd /opt && \
+        wget -q https://downloads.mysql.com/archives/get/p/10/file/mysql-connector-odbc-5.1.13-linux-glibc2.5-x86-64bit.tar.gz && \
+        tar xzf mysql-connector-odbc-5.1.13-linux-glibc2.5-x86-64bit.tar.gz && \
+        mysql-connector-odbc-5.1.13-linux-glibc2.5-x86-64bit/bin/myodbc-installer -d -a -n MySQL -t DRIVER=`ls /opt/mysql-*/lib/libmyodbc*w.so` && \
+        rm /opt/*.tar.gz
+
+    # we need an mssql odbc driver for the integration tests
+    RUN wget -q https://packages.microsoft.com/keys/microsoft.asc -O- | apt-key add - && \
+        wget -q https://packages.microsoft.com/config/debian/9/prod.list -O- > /etc/apt/sources.list.d/mssql-release.list && \
+        apt-get update && \
+        ACCEPT_EULA=Y apt-get install msodbcsql17 mssql-tools && \
+        odbcinst -i -d -f /opt/microsoft/msodbcsql17/etc/odbcinst.ini
+
+    # not used for the moment as Exasol is not tested here
+    # RUN cd /opt && \
+    #     wget -q https://www.exasol.com/support/secure/attachment/111075/EXASOL_ODBC-6.2.9.tar.gz && \
+    #     tar xzf EXASOL_ODBC-6.2.9.tar.gz && \
+    #     echo "\n[EXASOL]\nDriver=`ls /opt/EXA*/lib/linux/x86_64/libexaodbc-uo2214lv1.so`\nThreading=2\n" >> /etc/odbcinst.ini && \
+    #     rm /opt/*.tar.gz
+
+python:
+    ARG PYTHON_VERSION=3.8.6
+    ARG ARROW_VERSION_RULE="<2.0.0"
+    ARG NUMPY_VERSION_RULE=""
+    ARG CONDA_EXTRA=""
+    FROM +os-base
+
+    RUN mamba create -y -q -n turbodbc-dev \
+        c-compiler \
+        make \
+        ninja \
+        cmake \
+        coveralls \
+        gmock \
+        gtest \
+        cxx-compiler \
+        mock \
+        pytest \
+        pytest-cov \
+        python=${PYTHON_VERSION} \
+        unixodbc \
+        boost-cpp \
+        numpy$NUMPY_VERSION_RULE \
+        pyarrow$ARROW_VERSION_RULE \
+        pybind11 \
+        $CONDA_EXTRA
+
+    RUN echo "conda activate turbodbc-dev" >> ~/.bashrc
+    RUN echo 'export UNIXODBC_INCLUDE_DIR=$CONDA_PREFIX/include' >> ~/.bashrc
+
+    RUN cd /opt && \
+        wget -q https://github.com/pybind/pybind11/archive/v2.6.2.tar.gz && \
+        tar xvf v2.6.2.tar.gz
+
+build:
+    ARG PYTHON_VERSION=3.8.6
+    ARG ARROW_VERSION_RULE="<2.0.0"
+    ARG NUMPY_VERSION_RULE=""
+    ARG CONDA_EXTRA=""
+    FROM --build-arg PYTHON_VERSION="$PYTHON_VERSION" \
+        --build-arg ARROW_VERSION_RULE="$ARROW_VERSION_RULE" \
+        --build-arg NUMPY_VERSION_RULE="$NUMPY_VERSION_RULE" \
+        --build-arg CONDA_EXTRA="$CONDA_EXTRA" \
+        +python
+
+    COPY . /src
```

Comment:
This currently invalidates my local caches on every command execution, regardless of code changes or not. Haven't found the issue, yet, but I assume there is some log generated which invalidates this layer.

I assume that incremental builds are only possible if I use the interactive / manual setup you described

Reply:
Adding the offending file to `.dockerignore` should fix this?

Reply:
Is there any easy way to find the offender? Any docker build flag which prints info about the cache invalidation?

Reply:
I'm wondering what's best practice here. On could also think about a whitelisting with the dockerignore, e.g.

```
*
!cpp/**/*
!python/**/*
...
```

Reply:
I have to check if using a properly defined .earthignore https://docs.earthly.dev/earthfile/earthignore can improve this.

Reply:
Looks like ordinary `.dockerignore` doesn't work but we'll need an `.earthignore` https://docs.earthly.dev/earthfile/earthignore

and the magic directory to ignore is (git is not 100% necessary but e.g. VSCode fetches regularly in a background job which manipulates a file in `.git`)

```
.git*
.tmp-earthly*
```

Reply:
Sadly I think it does not help. Seems for me the COPY is always executed.
I can check with the earthly devs but I think we can live with this for now?